### PR TITLE
Changhiskhan/list versions

### LINKED ISSUE #511: list available versions
`lance.dataset(...).versions` should be a property that returns a list of versions in ascending order.
each version info should return the version number, the creation date, and any metadata associated with the version

### PR DESCRIPTION
**Pull request by changhiskhan**
closes #511

### CODE REVIEW

File: python/src/dataset.rs [hunk 1]
```diff
@@ -152,6 +152,34 @@ impl Dataset {
                 .map_err(|err| PyIOError::new_err(err.to_string()))?)
         })
     }
+
+    fn versions(self_: PyRef<'_, Self>) -> PyResult<Vec<PyObject>> {
+        let versions = self_
+            .list_versions()
+            .map_err(|err| PyValueError::new_err(err.to_string()))?;
```
Comment (eddyxu) [inline python/src/dataset.rs:159]: Make a  rust `Error` class for `lance/python`?

File: python/src/dataset.rs [hunk 2]
```diff
@@ -152,6 +152,34 @@ impl Dataset {
...
+                })
+                .collect::<Vec<_>>()
+                .into_iter()
+                .collect();
+            Ok(pyvers)
+        })
+    }
+}
+
+impl Dataset {
```
Comment (eddyxu) [inline python/src/dataset.rs:179]: is this a separated function from other implementations?  why do you need to have a separate `impl Dataset` here.

File: python/src/dataset.rs [hunk 3]
```diff
@@ -152,6 +152,34 @@ impl Dataset {
...
+    fn versions(self_: PyRef<'_, Self>) -> PyResult<Vec<PyObject>> {
+        let versions = self_
+            .list_versions()
+            .map_err(|err| PyValueError::new_err(err.to_string()))?;
+        Python::with_gil(|py| {
+            let pyvers: Vec<PyObject> = versions
+                .iter()
+                .map(|v| {
+                    let dict = PyDict::new(py);
+                    dict.set_item("version", v.version).unwrap();
```
Comment (eddyxu) [inline python/src/dataset.rs:165]: what does `dict.set_item()` return? `Result` or `Option`? Curious why it needs to unwrap().

File: python/src/dataset.rs (hunk 2 above)
Comment (changhiskhan) [reply python/src/dataset.rs:179]: so it's not exposed as pymethod

File: python/src/dataset.rs (hunk 3 above)
Comment (changhiskhan) [reply python/src/dataset.rs:165]: Result

**Pull request merged**